### ISSUE TITLE
camera and body in world coordinate system

### ISSUE THREAD

**Issue by davidpagnon** [pixelite1201/BEDLAM#44]
I can't reopen an issue and I'm not sure how the notification system works on closed ones, so I'm posting this as a new one. Please pardon me if this has already been seen. Here was my question:

I'm still kind of fuzzy about why and how the camera extrinsic parameters change from person to person on the same image. 
Why do we need to change both the position of the camera *(npz_params['cam_ext'])* and of the body *(npz_params['trans_world'])*?

In any case, since this is what I ultimately need, **is there a way to get the camera calibration in the world coordinate system, and idem for each SMPL-X mesh?**

> _Originally posted by @davidpagnon in https://github.com/pixelite1201/BEDLAM/issues/43#issuecomment-2035584229_

**Comment by davidpagnon** [comment]
Hi @pixelite1201 and co,

Now I'm thinking of giving up on using the npz data for the camera and body positions and locations, and rather use the *be_seq.csv* file. Once I'm sure the positions and orientations are good, I'll tackle pose, shape, and scale, but I suppose I will just have to retrieve the information from the npz file. **Is there a readme file or something that explains the formalism be_seq.csv uses?**

If I understand well, **Group** corresponds to the camera extrinsic parameters, and **Body** corresponds to the mesh ones.
Both Group and Body, X, Y, Z are in centimeters, and Yaw, Pitch, Roll in degrees.

Does it sound accurate? Do the location axes need to be rotated? In which order do the angles need to be taken?

**Comment by tpsmpi** [comment]
Always use the dedicated camera ground truth files (`seq_XXXXXX_camera.csv`) which we provide for retrieving camera extrinsics/intrinsics. There are many sequences in the dataset where the camera is moving. The `be_seq.csv` file does not provide camera ground truth for all types of shots in BEDLAM.

Camera ground truth .csv files use Unreal coordinate system to describe camera world space location and rotation.
+ unreal_X: forward, unreal_Y: right, unreal_Z:up
+ [cm]
+ Default camera faces along +unreal_X
+ Camera rotation is specified in Tait-Bryan angles: First yaw around world Z-axis, then pitch around new `local` Y-axis, then roll around new `local` X-axis. Note that this matches the order in the CSV file columns.
  + Yaw rotation is left-handed, pitch+roll rotations are right-handed

**Comment by tpsmpi** [comment]
See also: https://github.com/PerceivingSystems/bedlam_render/blob/main/unreal/render/unreal_coordinate_system.md

**Comment by davidpagnon** [comment]
Wow Unreal coordinate system is quite unusual! The weird transformations I was doing make a bit more sense now, thanks.
I'll experiment it next week and see if it fixes it all!
Best regards,

**Comment by davidpagnon** [comment]
Thank you @tpsmpi !

So I implemented the transformations you specified, and it seems like it is exactly equivalent to what I did above -- but at least, now it makes sense. However, there is still this little offset. 

I'm wondering if this might be an issue with the positioning of the body mesh. I enabled pose correctives, but it won't change location and rotation. There is this line I don't understand in the be_seq.csv file:
```csv
0,Comment,None,0,0,0,0,0,0,bodies_min=3;bodies_max=3;x_offset=650;y_offset=0.0;z_offset=0.0;x_min=-50;x_max=50;y_min=-250;y_max=250;yaw_min=0;yaw_max=360;cam_x_offset=10.0;cam_y_offset=10.0;cam_z_offset=5.0;cam_yaw_min=-3;cam_yaw_max=3;cam_pitch_min=-10;cam_pitch_max=3;cam_roll_min=-3;cam_roll_max=3;cam_config=cam_random_e
```
What does `x_offset=650;y_offset=0.0;z_offset=0.0` mean?
And `cam_x_offset=10.0;cam_y_offset=10.0;cam_z_offset=5.0`?

**Comment by tpsmpi** [comment]
Line index 0 ist just FYI comment about the randomization parameters.
The cameraroot parameters are used to setup the camera rig in Unreal for rendering.
See also: https://github.com/PerceivingSystems/bedlam_render/issues/10
But always use the dedicated camera ground truth files for anything camera related. They contain all you need.

**Comment by davidpagnon** [comment]
Nevermind, I figured it out! 
As I had to rotate my Blender camera so that it would face the +X direction, I faced some annoying brain teaser when trying to rotate around the new local axes, but it's all good now!

Thank you for taking the time to write [this very clear file](https://github.com/PerceivingSystems/bedlam_render/blob/main/unreal/render/unreal_coordinate_system.md) and for your help on this thread, I have been stuck on it for quite a while.
